keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,478 @@
+determine how up-to-date kubernetes watch caches are then serve reads from the
+watch cache when they are sufficiently up-to-date.
+
+## Motivation
+
+Serving reads from the watch cache is more performant and scalable than reading
+them from etcd, deserializing them, applying selectors, converting them to the
+desired version and then garbage collecting all the objects that were allocated
+during the whole process.
+
+We will need to measure the impact to performance and scalability, but we have
+enough data and experience from prior work with the watch cache to be confident
+there is significant scale/perf opportunity here, and we would like to introduce
+an alpha implementation.
+
+We expect the biggest gain to be from node-originating requests (e.g. kubelet
+listing pods scheduled on its node). For those requests, the size of the
+response is small (it fits a single page, assuming you won't make it extremely
+small), whereas the number of objects to process is proportional to cluster-size
+(so fairly big). For example, when kubelets requests pods schedule against it in
+a 5k node cluster with 30pods/node, the kube-apiserver must list the 150k pods
+from etcd and then filter that list down to the list of 30 pods that the kubelet
+actually need. This must occur for each list request from each of the 5k
+kubelets. If served from watch cache, this same request can be served by simply
+filtering out the 30 pods each kubelet needs from the data in the cache.
+
+In addition to the improvements to scale and performance, we aim to resolve a
+specific problem. The long standing "stale read" issue
+(https://github.com/kubernetes/kubernetes/issues/59848) remains open because
+reflectors default to resourceVersion=”0” for their initial list requests. If
+the reflectors instead use a consistent read for their initial list request,
+they could not "going back in time" when components are restarted and this issue
+would be solved. "Going back in time" can curently happen if the initial list
+request is served from a stale watch cache with data much older than the
+reflector has previously observed or if the api-server or etcd are partitioned.
+
+We have held off on switching reflectors to using consistent read for the
+initial list, even though we know it is more correct, due to concerns with the
+impact on large scale use cases. But if we serve consistent reads from cache,
+there would be very little difference in scalability to how the kube-apiserver
+serves the resourceVersion="0" list requests from reflectors today.
+
+### Goals
+
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scalability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+### Non-Goals
+
+<<[UNRESOLVED @deads]>>
+- Avoid allowing true quorum reads. We should think carefully about this, see: https://github.com/kubernetes/enhancements/pull/1404#discussion_r381528406
```

Comment:
i feel the "right and traditional" solution for this is to enable v system leasing in the apiserver etcd pkg. here is the description on how it works: http://web.stanford.edu/class/cs240/readings/89-leases.pdf. we already implemented it here: https://github.com/etcd-io/etcd/pull/8341/files. we can port the code into the apiserver side if we want. but this is a pretty heavy :P